Add support for Workstation 17.6.3

philipl opened the pull request:
Very very minor changes to some of the RHEL support macros. This works with 6.13.x and 6.14.x kernels.

Edit: Updated with support for 6.15.x and 6.16.x

Edit: 17.6.4 is identical.

Edit: Kernel 6.17.x doesn't require any additional changes (surprise!)

mkubecek committed:
import Workstation 12.5.5 module sources

mkubecek committed:
import Workstation 12.5.6 module sources

mkubecek committed:
import Workstation 12.5.7 module sources

mkubecek committed:
import Workstation 14.0.0 module sources

mkubecek committed:
vmmon: quick workaround for objtool warnings

As discussed in

  https://bugzilla.suse.com/show_bug.cgi?id=1059674

the reason for multiple objtool warnings is the fact that vmmon module
defines its own Panic() function which never returns. While it is marked as
such which is used by the compiler for optimization, there is no way to
find this from object file.

While this seems innocuous, it might result in problems with unwinder
later. The quickest way around is to replace vmmon's own Panic() with
standard kernel panic() until a cleaner solution is found.

mkubecek committed:
import Workstation 14.1.0 module sources

mkubecek committed:
import Workstation 14.1.1 module sources

mkubecek committed:
Merge branch 'master' into workstation

mkubecek committed:
Merge branch 'master' into workstation

mkubecek committed:
import Workstation 14.1.2 module sources

theobch committed:
modules: remove .cache.mk on make clean

Remove .cache.mk inside vmmon-only and vmnet-only when executing
"make clean". This file can cause issues when upgrading gcc as make will
still look for includes inside older gcc includes directory.

File .cache.mk with cache of generated variables was created by build since
kernel v4.15-rc1, commit 3298b690b21c ("kbuild: Add a cache for generated
variables") until the feature was removed in v4.18-rc1, commit e08d6de4e532
("kbuild: remove kbuild cache").

mkubecek committed:
Merge branch 'master' into workstation

mkubecek committed:
import Workstation 14.1.3 module sources

mkubecek committed:
import Workstation 15.0.0 module sources

mkubecek committed:
import Workstation 15.0.1 module sources

mkubecek committed:
import Workstation 15.1.0 module sources

mkubecek committed:
Merge branch 'master' into workstation

mkubecek committed:
import Workstation 15.5.0 module sources

mkubecek committed:
import Workstation 15.5.1 module sources

mkubecek committed:
import Workstation 15.5.2 module sources

mkubecek committed:
vmnet: convert VNetBridge debugging code to ktime_get_ns()

The do_gettimeofday() helper was removed by commit e4b92b108c6c
("timekeeping: remove obsolete time accessors") in v5.0-rc1 and since
commit c766d1472c70 ("y2038: hide timeval/timespec/itimerval/itimerspec
types") in v5.6-rc3, struct timeval should no longer be used in kernel
code.

Convert the do_gettimeofday() relics in VNetBridge (which are only compiled
with LOGLEVEL >= 4) completely to ktime based interface.

mkubecek committed:
import Workstation 15.5.5 module sources

mkubecek committed:
vmmon: include <asm/irq_vectors.h> explicitly

After mainline commit 13c01139b171 ("x86/headers: Remove APIC headers from
<asm/smp.h>") in 5.9-rc1, APIC headers are no longer included via
<asm/smp.h> so that linux/hostif.c will use incorrect fallback definitions
of SPURIOUS_APIC_VECTOR, POSTED_INTR_VECTOR and ERROR_APIC_VECTOR even if
built against kernel where these are defined.

Include <asm/irq_vectors.h> in linux/hostif.c explicitly to avoid that.

mkubecek committed:
import Workstation 16.0.0 module sources

mkubecek committed:
vmmon: rename GDT_SIZE macro to avoid collision

When building against kernel 4.12 and newer, macro name GDT_SIZE used in
(vmmon) include/segs.h collides with macro defined in (kernel source)
arch/x86/include/asm/segment.h, resulting in its redefinition.

To prevent potential problems, rename vmmon's GDT_SIZE to VMMON_GDT_SIZE
and GDT_LIMIT to VMMON_GDT_LIMIT. (There is no GDT_LIMIT in mainline kernel
source but let's be consistent.)

mkubecek committed:
import Workstation 16.1.0 module sources

mkubecek committed:
import Workstation 16.2.0 module sources

mkubecek committed:
import Workstation 16.2.3 module sources

mkubecek committed:
Merge branch 'master' into workstation

mkubecek committed:
vmnet: fix wireless device detection

Mainline commit c304eddcecfe ("net: wrap the wireless pointers in struct
net_device in an ifdef") in 5.19-rc1 makess ieee80211_ptr member present in
struct net_device only if CONFIG_CFG80211 is enabled. Workstation 17.0.0
checks if CONFIG_CFG80211 is enabled but only when CONFIG_WIRELESS_EXT is
not. Thus a build against a kernel with CONFIG_WIRELESS_EXT enabled and
CONFIG_CFG80211 disabled still fails. Also, the newly introduced version
check is pointless, there is no point checking dev->ieee80211_ptr with
CONFIG_CFG80211 disabled (fortunately it will be null anyway).

Rewrite the check in VNetBridgeIsDeviceWireless() to check each of the two
pointers only if it actually exists.

mkubecek committed:
import Workstation 16.2.4 module sources

mkubecek committed:
Merge branch 'master' into workstation

mkubecek committed:
Merge branch 'master' into workstation

mkubecek committed:
import Workstation 17.0.0 module sources

mkubecek committed:
modules: use standard constant definitions if available

The PCI_VENDOR_ID_VMWARE macro is defined in mainline pci_ids.h since
commit 94e57fea6202 ("PCI: Move PCI_VENDOR_ID_VMWARE to pci_ids.h") in
v3.18-rc1.

The PCI_DEVICE_ID_VMWARE_VMXNET3 macro is defined in mainline pci_ids.h
since commit b1226c7db1d9 ("vmxnet3: Move PCI Id to pci_ids.h") in
v4.10-rc1.

The MSR_MISC_FEATURES_ENABLES_CPUID_FAULT macro is defined in mainline
since commit e9ea1e7f53b8 ("x86/arch_prctl: Add ARCH_[GET|SET]_CPUID") in
v4.12-rc1.

The CR3_PCID_MASK is defined in mainline asm/processor-flags.h since commit
6c690ee1039b ("x86/mm: Split read_cr3() into read_cr3_pa() and
__read_cr3()") in v4.13-rc1.

mkubecek committed:
vmnet: fix csum_and_copy_to_user workaround for kernel >= 5.19

Mainline commit 6308499b5e99 ("net: unexport csum_and_copy_{from,to}_user")
in 5.19-rc1 unexports csum_and_copy_to_user as no in-tree module is using
it. A clean solution would probably be rewriting the code to use iovec
iterator as csum_and_copy_to_iter() is still exported (or perhaps
skb_copy_and_csum_datagram() might be used instead). Anything like this
would be way too intrusive so it would have to wait for VMware developers.

Workstation 17.0.0 handles this with a call to csum_partial_copy_nocheck()
inside a user_access_begin()/user_access_end() block which lets the build
succeed but objtool still warns about a call to csum_partial_copy_nocheck()
with UACCESS enabled. Based on the reasoning in commit message of mainline
commit ea24213d8088 ("objtool: Add UACCESS validation"), this workaround
does indeed seem questionable.

Use the older workaround combining csum_partial() with copy_to_user() like
in workstation-16.2.4 branch instead. This will be less efficient but
hopefully the performace hit will not be noticeable.

mkubecek committed:
vmnet: hack around SLE15-SP5 backport of netdevice::dev_addr changes

SLE15-SP5 backported mainline commit adeef3e32146 ("net: constify
netdev->dev_addr") from 5.17-rc1 into their "5.14" kernel. Add an extra
hack to the version check to fix SLE15-SP5 build.

mkubecek committed:
vmmon: use standard ASM_ENDBR macro

Some cross page functions need an explicit endbr64 instruction as they are
indirect branch targets but are not recognized as such. VMware 17 uses home
cooked ENDBR macro rather than standard ASM_ENDBR defined in kernel.

Use ASM_ENDBR instead and define it as empty if not available (kernel
before 5.18-rc1) so that we do not generate useless endbr64 instructions
when building against kernel which does not support IBT or has it disabled.

mkubecek committed:
vmnet: use explicit module_init() and module_exit() (#187)

While vmmon module already uses explicit module_init() and module_exit()
for its init and cleanup function, vmnet relies on traditional magic names
init_module() and cleanup_module(). Apparently this has an unfortunate side
effect that the two functions are not identified as indirect call targets
by objdump and they get "sealed" when the module is built against and
loaded into an IBT enabled kernel.

Starting with 6.3-rc1, objtool is going to warn about this issue,
indicating that the legacy module initialization is deprecated and
module_init() and module_exit() macros should be used instead so do that
for vmnet as well.

mkubecek committed:
import Workstation 17.0.1 module sources

mkubecek committed:
vmnet: use void as argument list in function definitions (#188)

Two functions use "foo()" rather than proper "foo(void)" in their
definitions and as reported, clang compiler treats it as an error.

While at it, also mark VNetFreeInterfaceList() static to make its
definition match the declaration.

mkubecek committed:
modules: make COMPAT_LINUX_VERSION_CHECK_LT less tricky

The definition of COMPAT_LINUX_VERSION_CHECK_LT() macro lacks surrounding
paretheses so that negated tests like !COMPAT_LINUX_VERSION_CHECK_LT(...)
expand to something completely different. This could be worked around
easily by adding parentheses to each place the macro is used in an
expression but it makes much more sense to fix the macro definition so that
the macro does not serve as a trap.

mkubecek committed:
import Workstation 17.5.0 module sources

philipl committed:
import Workstation 17.6.0 module sources

philipl committed:
Add dkms support

This also requires tweaking the Makefiles to allow for building for non-active
kernels

philipl committed:
import Workstation 17.6.1 module sources

philipl committed:
import Workstation 17.6.3 module sources

noahfriedman commented:
I still get the compilation error

vmmon.o: warning: objtool: CrossPage_CodePage+0x1d6: corrupt function pointer reference

With the 6.13.9-200.fc41.x86_64 kernel. In fact I've been getting that corrupt function pointer reference problem with earlier kernels as well but I didn't track them.  Can anyone else reproduce this using this branch?

philipl committed:
vmmon: adjust for del_timer_sync renamed to timer_delete_sync

philipl committed:
vmmon, vmnet: switch from EXTRA_CFLAGS to ccflags-y

I'm not sure how far back this change is compatible. Maybe someone will
let me know.

leandromqrs commented:
Worked great in Fedora 42 with kernel 6.15.3-200.fc42.x86_64 without errors! Thanks!

zDEFz commented:
Hello. I probably did something wrong, but when I apply this, my VM network will stop working.
Archlinux latest 6.15.4-arch2-1.1

powellnorma commented:
@zDEFz Does VM network work fine when you use https://github.com/mkubecek/vmware-host-modules/issues/306#issuecomment-2843789954 ? 

Edit:
For me the modules in the linked .tar.gz file worked, after `sudo systemctl restart vmware`

zDEFz commented:
> @zDEFz Does VM network work fine when you use [#306 (comment)](https://github.com/mkubecek/vmware-host-modules/issues/306#issuecomment-2843789954) ?
> 
> Edit: For me the modules in the linked .tar.gz file worked, after `sudo systemctl restart vmware`

Sorry, but I think this is a displaced question.
I won't blindly use the .tar.gz
I did once, and that was an absolute exception.
Expecting a PR, then will test.

powellnorma commented:
Well, I asked so that we can better pinpoint if there is an issue in this PR, or in your setup in general.
 
Also about reviewability / transparency about made changes, I feel like a PR with 76k LOC change (like this one) is not much better, optimally we'd have a PR that just shows the diff (what changes needed to be done so it compiles) or am I missing something?

powellnorma commented:
> I feel like a PR with 76k LOC change (like this one) is not much better

I just see, that @philipl did something that leads to the actual changes he made to be shown at the bottom of the commit history. 

I made a branch from the changes of @jeamieofqidan for 6.15 here:
https://github.com/powellnorma/vmware-host-modules/commits/workstation-17.6.3_jeamieofqidan/

zDEFz commented:
> > I feel like a PR with 76k LOC change (like this one) is not much better
> 
> I just see, that @philipl did something that leads to the actual changes he made to be shown at the bottom of the commit history.
> 
> I made a branch from the changes of @jeamieofqidan for 6.15 here: https://github.com/powellnorma/vmware-host-modules/commits/workstation-17.6.3_jeamieofqidan/

How does that differ from the version that works from kernel 6.16?

philipl commented:
This PR is structured to create a branch for 17.6 that follows the pattern used for creating branches for older releases. As a result it's not going to be just the changes required to fix compilation. If you look at the source branch for the PR itself, the history is clear.

powellnorma commented:
> How does that differ from the version that works from kernel 6.16?

You'd need to take the tgz for 6.16 and overwrite all files, then commit the changes. This way you'd get a commit which contains the diff, to see what jeamieofqidan changed for 6.16 compatibility. I might do this later if I find the time for it.

> This PR is structured to create a branch for 17.6 that follows the pattern used for creating branches for older releases. As a result it's not going to be just the changes required to fix compilation

But why the latest commits show up as first commits? It does not make intuitive sense. I think it should be possible to add your newer commits afterward, not before. Why is it done this way?

philipl commented:
> But why the latest commits show up as first commits? It does not make intuitive sense. I think it should be possible to add your newer commits afterward, not before. Why is it done this way?

I don't know what you mean. github decides how to show the commits in a PR, but they are in ascending order on my screen, with the newest commits at the bottom of the list.

zDEFz commented:
> > But why the latest commits show up as first commits? It does not make intuitive sense. I think it should be possible to add your newer commits afterward, not before. Why is it done this way?
> 
> I don't know what you mean. github decides how to show the commits in a PR, but they are in ascending order on my screen, with the newest commits at the bottom of the list.

not for me. what do you see?
<img width="1917" height="732" alt="image" src="https://github.com/user-attachments/assets/a2d8ad86-2db3-4611-8321-2b53f91b2855" />

philipl commented:
That screenshot isn't from this PR's commit list, so I don't know what you're looking at.

zDEFz commented:
> That screenshot isn't from this PR's commit list, so I don't know what you're looking at.

It's from the link that was posted https://github.com/powellnorma/vmware-host-modules/commits/workstation-17.6.3_jeamieofqidan/

What link did you visit?

philipl commented:
The commits for this PR. That we are adding comments to.

The link you are referring to is the commit list for a branch, and github shows a branch's commit history in descending order with the most recent first. That's normal git behaviour.

powellnorma commented:
> > This PR is structured to create a branch for 17.6 that follows the pattern used for creating branches for older releases. As a result it's not going to be just the changes required to fix compilation

> But why the latest commits show up as first commits? 

What I meant were the oldest commits, I got confused why they were shown at the beginning of the PR commit history.
But now I see it is probably just GH using different ordering for "branch commit history" and "PR commit history".

For new PRs, GH seems to by default base them off from the repos master branch.
I think it would make sense to make it such that master points to the latest commit/branch.
This way the PR should show nicer. But that's not something @philipl can influence, ofc. @mkubecek would need to decide on that.

What one can do as contributor though, is to chose as base branch the latest available one, instead of master.
Example: https://github.com/mkubecek/vmware-host-modules/pull/316

philipl committed:
vmon: replace rdmsrl_safe with rdmsrq_safe for >= 6.16

philipl commented:
Added the additional patch needed for >= 6.16

zDEFz commented:
Updated - works and tested. Thanks @philipl https://aur.archlinux.org/packages/vmware-host-modules-dkms-fix-git
